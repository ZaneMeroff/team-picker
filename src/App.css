* {
  margin: 0px;
  padding: 0px;
  color: #FFF;
  font-family: 'Play', sans-serif;
}

.app-title {
  font-size: 200%;
  margin-top: 20px;
}

.main-outer-container {
  display: flex;
  align-items: center;
  justify-content: center;
  height: auto;
  width: 100vw;
  background-color: #1e1e2e;
}

.main-container {
  height: 100vh;
  width: 360px;
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: #034843;
}

.view-container {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.nav-button-container {
  height: 50px;
  width: 250px;
  display: flex;
  justify-content: space-between;
  margin: 15px 0px;
}

.arrow-button {
  height: 50px;
  width: 50px;
  border-radius: 8px;
  font-size: 150%;
  background-color: #3e7cee;
  border: 2px solid #FFF;
  color: #FFF;
}

.arrow-button:hover {
  cursor: pointer;
}

.error-message {
  position: absolute;
  top: 130px;
  color: #ffad00;
}

@media only screen and (max-width: 490px) {

  .main-container {
    width: 100%;
    height: 100vh;
  }

}
