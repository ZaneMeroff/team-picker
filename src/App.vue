<template>
  <div id='main-outer-container'>
    <div class='main-container'>
      <h1>Random Team Generator</h1>

      <div v-if='currentStep === 1'>
        <NumSelector label='teams' :number='numberOfTeams' :updateNumber='updateNumber'/>
        <NumSelector label='players' :number='playerNames.length' :updateNumber='updateNumber'/>
      </div>

      <div v-if='currentStep === 2'>
        <button class='back-button' @click="nextScreen('back')">back</button>
        <div v-for='(player, index) in playerNames' :key='index' >
          <PlayerNameInput :id='index' :updateNames='updateNames'/>
        </div>
      </div>

      <button class='next-button' @click="nextScreen('next')">next</button>
    </div>
  </div>
</template>

<script>
  import NumSelector from './components/NumSelector/NumSelector.vue';
  import PlayerNameInput from './components/PlayerNameInput/PlayerNameInput.vue';
  export default {
    name: 'App',
    components: {
      NumSelector,
      PlayerNameInput
    },
    data() {
      return {
        numberOfTeams: 2,
        playerNames: ['',''],
        currentStep: 1
      }
    },
    methods: {
      nextScreen(nextOrBack) {
        if (nextOrBack === 'next') {
          this.currentStep++
        } else if (nextOrBack === 'back') {
          this.currentStep--
        }
      },
      updateNumber(action) {
        const teamKey = {
          'teamsUp': 'this.numberOfTeams++',
          'teamsDown': 'this.numberOfTeams--'
        }
        eval(teamKey[action])
        this.updatePlayers(action)
      },
      updatePlayers(action) {
        if (action === 'playersUp') {
          this.playerNames.push('')
        } else if (action === 'playersDown') {
          this.playerNames.pop()
        }
      },
      updateNames(name, index) {
        this.playerNames[index] = name;
      }
    },
    watch: {
      numberOfTeams() {
        if (this.numberOfTeams < 2) {this.numberOfTeams = 2}
      },
      playerNames() {
        if (this.playerNames.length < 2) {this.playerNames.push('')}
      }
    }
  }
  </script>

  <style>
    * {
      margin: 0px;
      padding: 0px;
    }
    #main-outer-container {
      display: flex;
      align-items: center;
      justify-content: center;
      height: 100vh;
      width: 100vw;
    }
    .main-container {
      height: 500px;
      width: 500px;
      display: flex;
      flex-direction: column;
      align-items: center;
      border: 2px solid #000;
    }
    .next-button {
      height: 50px;
      width: 80px;
      margin: 20px;
    }
    .next-button:hover {
      cursor: pointer;
    }
    .back-button {
      height: 30px;
      width: 50px;
      margin-left: -100px;
      margin-top: 10px;
    }
</style>
