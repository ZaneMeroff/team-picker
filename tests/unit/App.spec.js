import { shallowMount } from '@vue/test-utils'
import App from '@/App.vue'

describe('App.vue', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(App)
  })

  it('component render matches snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('component has correct default data values', () => {
    expect(wrapper.vm.$data.numberOfTeams).toEqual(2)
    expect(wrapper.vm.$data.playerNames).toEqual(['',''])
    expect(wrapper.vm.$data.teamObjects).toEqual([])
    expect(wrapper.vm.$data.currentStep).toEqual(1)
    expect(wrapper.vm.$data.playerCountErrorMessage).toEqual(false)
    expect(wrapper.vm.$data.nameInputsErrorMessage).toEqual(false)
  })

  it('buildTeamObjects creates correct number of teams', () => {
    expect(wrapper.vm.$data.teamObjects.length).toEqual(0)
    wrapper.vm.buildTeamObjects(2)
    expect(wrapper.vm.$data.teamObjects.length).toEqual(2)
  })

  it('distributeOddNumOfPlayers adds additional player appropriately', () => {
    let mockName = ['Tommy']
    let mockTeamObjects = [
      { teamID: 1, names: ['Chuckie', 'Angelica'] },
      { teamID: 2, names: ['Phil', 'Lil'] }
    ]
    let expected = [
      { teamID: 1, names: ['Chuckie', 'Angelica', 'Tommy'] },
      { teamID: 2, names: ['Phil', 'Lil'] }
    ]
    wrapper.vm.$data.teamObjects = mockTeamObjects
    wrapper.vm.distributeOddNumOfPlayers(mockName)
    expect(wrapper.vm.$data.teamObjects).toEqual(expected)
  })

  it('generateTeams correctly distributes players', () => {
    let mockNames = ['Tommy', 'Chuckie', 'Phil', 'Lil']
    let mockNumOfTeams = 2
    wrapper.vm.nextScreen = jest.fn()
    expect(wrapper.vm.$data.teamObjects.length).toEqual(0)
    wrapper.vm.generateTeams(mockNames, mockNumOfTeams)
    expect(wrapper.vm.$data.teamObjects.length).toEqual(2)
    expect(wrapper.vm.nextScreen).toHaveBeenCalledWith('next')
  })

  describe('validatePlayersExceedTeams', () => {

    it('if numberOfTeams exceeds number of players, error message is true', () => {
      let mockPlayerNames = ['Tommy', 'Chuckie']
      wrapper.vm.$data.numberOfTeams = 3
      wrapper.vm.$data.playerNames = mockPlayerNames
      expect(wrapper.vm.playerCountErrorMessage).toEqual(false)
      wrapper.vm.validatePlayersExceedTeams()
      expect(wrapper.vm.playerCountErrorMessage).toEqual(true)
    })

    it('if numberOfTeams does not exceed number of players, error message is false', () => {
      let mockPlayerNames = ['Tommy', 'Chuckie']
      wrapper.vm.$data.numberOfTeams = 2
      wrapper.vm.$data.playerNames = mockPlayerNames
      expect(wrapper.vm.playerCountErrorMessage).toEqual(false)
      wrapper.vm.validatePlayersExceedTeams()
      expect(wrapper.vm.playerCountErrorMessage).toEqual(false)
    })

  })

  describe('validatePlayerNameInputs', () => {

    it('if playerNames contains an empty string, error message is true', () => {
      let mockPlayerNames = ['Tommy', '', 'Chuckie']
      wrapper.vm.$data.playerNames = mockPlayerNames
      expect(wrapper.vm.$data.nameInputsErrorMessage).toEqual(false)
      wrapper.vm.validatePlayerNameInputs()
      expect(wrapper.vm.$data.nameInputsErrorMessage).toEqual(true)
    })

    it('if playerNames does not contain an empty string, error message is false ', () => {
      let mockPlayerNames = ['Tommy', 'Chuckie', 'Phil']
      wrapper.vm.$data.playerNames = mockPlayerNames
      expect(wrapper.vm.$data.nameInputsErrorMessage).toEqual(false)
      wrapper.vm.validatePlayerNameInputs()
      expect(wrapper.vm.$data.nameInputsErrorMessage).toEqual(false)
    })

    it('if playerNames does not contain an empty string, generateTeams is called', () => {
      let mockPlayerNames = ['Tommy', 'Chuckie', 'Phil']
      wrapper.vm.generateTeams = jest.fn()
      wrapper.vm.$data.playerNames = mockPlayerNames
      wrapper.vm.validatePlayerNameInputs()
      expect(wrapper.vm.generateTeams).toHaveBeenCalled()
    })

  })

  describe('nextScreen', () => {

    it('if nextScreen is called with next, currentStep is incremented by 1', () => {
      expect(wrapper.vm.$data.currentStep).toEqual(1)
      wrapper.vm.nextScreen('next')
      expect(wrapper.vm.$data.currentStep).toEqual(2)
    })

    it('if nextScreen is called with back, error message is false', () => {
      wrapper.vm.$data.nameInputsErrorMessage = true
      expect(wrapper.vm.$data.nameInputsErrorMessage).toEqual(true)
      wrapper.vm.nextScreen('back')
      expect(wrapper.vm.$data.nameInputsErrorMessage).toEqual(false)
    })

    it('if nextScreen is called with back, currentStep is decremented by 1', () => {
      wrapper.vm.$data.currentStep = 2
      expect(wrapper.vm.$data.currentStep).toEqual(2)
      wrapper.vm.nextScreen('back')
      expect(wrapper.vm.$data.currentStep).toEqual(1)
    })

  })

  describe('updateNumber', () => {

    it('if updateNumber is called with teamsUp, numberOfTeams is incremented by 1', () => {
      let mockAction = 'teamsUp'
      wrapper.vm.$data.numberOfTeams = 4
      expect(wrapper.vm.$data.numberOfTeams).toEqual(4)
      wrapper.vm.updateNumber(mockAction)
      expect(wrapper.vm.$data.numberOfTeams).toEqual(5)
    })

    it('if updateNumber is called with teamsDown, numberOfTeams is decremented by 1', () => {
      let mockAction = 'teamsDown'
      wrapper.vm.$data.numberOfTeams = 4
      expect(wrapper.vm.$data.numberOfTeams).toEqual(4)
      wrapper.vm.updateNumber(mockAction)
      expect(wrapper.vm.$data.numberOfTeams).toEqual(3)
    })

    it('if updateNumber is called with teamsUp, updatePlayers is called with teamsUp', () => {
      let mockAction = 'teamsUp'
      let expected = 'teamsUp'
      wrapper.vm.updatePlayers = jest.fn()
      wrapper.vm.updateNumber(mockAction)
      expect(wrapper.vm.updatePlayers).toHaveBeenCalledWith(expected)
    })

    it('if updateNumber is called with teamsDown, updatePlayers is called with teamsDown', () => {
      let mockAction = 'teamsDown'
      let expected = 'teamsDown'
      wrapper.vm.updatePlayers = jest.fn()
      wrapper.vm.updateNumber(mockAction)
      expect(wrapper.vm.updatePlayers).toHaveBeenCalledWith(expected)
    })

  })

  describe('updatePlayers', () => {

    it('if updatePlayers is called with playersUp, playerNames.length increases by 1', () => {
      let mockPlayerNames = ['Tommy', 'Chuckie', 'Phil']
      wrapper.vm.$data.playerNames = mockPlayerNames
      expect(wrapper.vm.$data.playerNames.length).toEqual(3)
      wrapper.vm.updatePlayers('playersUp')
      expect(wrapper.vm.$data.playerNames.length).toEqual(4)
    })

    it('if updatePlayers is called with playersDown, playerNames.length decreases by 1', () => {
      let mockPlayerNames = ['Tommy', 'Chuckie', 'Phil']
      wrapper.vm.$data.playerNames = mockPlayerNames
      expect(wrapper.vm.$data.playerNames.length).toEqual(3)
      wrapper.vm.updatePlayers('playersDown')
      expect(wrapper.vm.$data.playerNames.length).toEqual(2)
    })

    it('if updatePlayers is called with potato, playerNames.length stays the same', () => {
      let mockPlayerNames = ['Tommy', 'Chuckie', 'Phil']
      wrapper.vm.$data.playerNames = mockPlayerNames
      expect(wrapper.vm.$data.playerNames.length).toEqual(3)
      wrapper.vm.updatePlayers('potato')
      expect(wrapper.vm.$data.playerNames.length).toEqual(3)
    })

  })

  it('updateNames correctly finds and updates a player name based on index', () => {
    let mockPlayerNames = ['', '', '']
    let expected = ['Tommy', '', '']
    wrapper.vm.$data.playerNames = mockPlayerNames
    expect(wrapper.vm.$data.playerNames).toEqual(mockPlayerNames)
    wrapper.vm.updateNames('Tommy', 0)
    expect(wrapper.vm.$data.playerNames).toEqual(expected)
  })

})
