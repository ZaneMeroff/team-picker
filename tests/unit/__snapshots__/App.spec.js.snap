// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue component render matches snapshot 1`] = `
<div class="main-outer-container">
  <div class="main-container">
    <h1 class="app-title">Team Generator</h1>
    <div class="view-container">
      <div class="nav-button-container">
        <div></div> <button class="arrow-button">â†’</button>
        <!---->
      </div>
      <numselector-stub label="teams" number="2" updatenumber="function () { [native code] }"></numselector-stub>
      <numselector-stub label="players" number="2" updatenumber="function () { [native code] }"></numselector-stub>
    </div>
    <!---->
    <!---->
  </div>
</div>
`;
